src/NuGet.Core/NuGet.PackageManagement/Audit/AuditChecker.cs

```diff
@@ -55,18 +71,44 @@ public async Task CheckPackageVulnerabilitiesAsync(CancellationToken cancellatio
 
             if (allVulnerabilityData is null || !IsAnyVulnerabilityDataFound(allVulnerabilityData.KnownVulnerabilities))
             {
-                return;
+                return new AuditCheckResult(Array.Empty<ILogMessage>())
+                {
+                    DownloadDurationInSeconds = downloadDurationInSeconds,
+                    SourcesWithVulnerabilities = sourceWithVulnerabilityCount,
+                };
             }
 
-            Dictionary<PackageIdentity, PackageAuditInfo>? packagesWithKnownVulnerabilities =
-                FindPackagesWithKnownVulnerabilities(allVulnerabilityData.KnownVulnerabilities!,
-                                                    _packages,
-                                                    _minSeverity);
-            if (packagesWithKnownVulnerabilities is not null)
+            stopwatch.Restart();
+            Dictionary<PackageIdentity, PackageAuditInfo>? packagesWithKnownVulnerabilities = FindPackagesWithKnownVulnerabilities(allVulnerabilityData.KnownVulnerabilities!, packages);
+            int Sev0Matches = 0, Sev1Matches = 0, Sev2Matches = 0, Sev3Matches = 0, InvalidSevMatches = 0;
+
+            List<PackageIdentity> packagesWithReportedAdvisories = new(packagesWithKnownVulnerabilities?.Count ?? 0);
+
+            IReadOnlyList<ILogMessage> warnings = packagesWithKnownVulnerabilities is not null ?
+                CreateWarnings(packagesWithKnownVulnerabilities, auditSettings, ref Sev0Matches, ref Sev1Matches, ref Sev2Matches, ref Sev3Matches, ref InvalidSevMatches, ref packagesWithReportedAdvisories) :
+                Array.Empty<ILogMessage>();
+
+            foreach (var warning in warnings)
```

Comment:
```suggestion
            foreach (var warning in warnings.NoAllocEnumerate())
```

`warnings` is `IReadOnlyList<T>`, hence `foreach` will do a heap alloc by default.